leiyu/obmc-utils/upload_and_update.py

```diff
@@ -0,0 +1,222 @@
+from time import sleep
+
+
+def get_tftp_ip(tftp):
+    if '@' in tftp:
+        ip = tftp.split('@')[1].split(':')[0]
+    else:
+        ip = tftp.split(':')[0]
+    return ip
+
+
+def get_filename(tarball):
+    return os.path.basename(tarball)
+
+def get_server_filename(tftp, tarball):
+    if '@' in tftp:
+        user = tftp.split('@')[0]
+    else:
+        import getpass
+        user = getpass.getuser()
+    return get_filename(tarball) + "_" + user
+
+
+def checkBmcAlive(bmc):
+    cmds = ['ping', '-c', '1', bmc]
+    try:
+        check_call(cmds, stdout=FNULL, stderr=FNULL)
+    except CalledProcessError:
+        return False
+    else:
+        return True
+
+
+def login(bmc):
+    url = 'https://%s/login' % bmc
+    cmds = ['curl', '-s', '-c', 'cjar', '-k', '-X', 'POST', '-H',
+            'Content-Type: application/json', '-d',
+            '{"data": [ "root", "0penBmc"]}', url]
+    try:
+        check_call(cmds, stdout=FNULL, stderr=FNULL)
+    except CalledProcessError:
+        return False
+    else:
+        return True
+
+
+def prepare(bmc):
+    url = 'https://%s/org/openbmc/control/flash/bmc/action/prepareForUpdate' % bmc
+    cmds = ['curl', '-s', '-b', 'cjar', '-k', '-X', 'POST', '-H',
+            'Content-Type: application/json', '-d',
+            '{"data": []}', url]
+    check_call(cmds, stdout=FNULL, stderr=FNULL)
+
+
+def preserveNetwork(bmc):
+    url = 'https://%s/org/openbmc/control/flash/bmc/attr/preserve_network_settings' % bmc
+    cmds = ['curl', '-s', '-b', 'cjar', '-k', '-X', 'PUT', '-H',
+            'Content-Type: application/json', '-d',
+            '{"data": 1}', url]
+    check_call(cmds, stdout=FNULL, stderr=FNULL)
+
+
+def updateViaTFTP(tftp, tarball, bmc):
+    tftp_ip = get_tftp_ip(tftp)
+    serverfile = get_server_filename(tftp, tarball)
+    url = 'https://%s/org/openbmc/control/flash/bmc/action/updateViaTftp' % bmc
+    cmds = ['curl', '-s', '-b', 'cjar', '-k', '-X', 'POST', '-H',
+            'Content-Type: application/json', '-d',
+            '{"data": ["%s", "%s"]}' % (tftp_ip, serverfile), url]
+    check_call(cmds, stdout=FNULL, stderr=FNULL)
+
+
+def applyUpdate(bmc):
+    url = 'https://%s/org/openbmc/control/flash/bmc/action/Apply' % bmc
+    cmds = ['curl', '-s', '-b', 'cjar', '-k', '-X', 'POST', '-H',
+            'Content-Type: application/json', '-d',
+            '{"data": []}', url]
+    check_call(cmds, stdout=FNULL, stderr=FNULL)
+
+
+def getProgress(bmc):
+    url = 'https://%s/org/openbmc/control/flash/bmc/action/GetUpdateProgress' % bmc
+    cmds = ['curl', '-s', '-b', 'cjar', '-k', '-X', 'POST', '-H',
+            'Content-Type: application/json', '-d',
+            '{"data": []}', url]
+    try:
+        output = subprocess.check_output(cmds, stderr=FNULL)
+    except CalledProcessError as e:
+        # Sometimes curl fails with timeout error, let's ignore it
+        return ''
+    if FNULL is None:  # Do not print log when FNULL is devnull
```

Comment:
Can't we control the subprocess' stdout/stderr? Wouldn't that be better than buffering all the output?